File: draft_design_pattern/decorator.py
# decoratorパターンは、オブジェクトをdecoratorクラスのオブジェクトにラップすることで、
# 実行時にオブジェクトの振舞いを動的に変更できるようにします
from abc import ABC, abstractmethod


class Coffee(ABC):
    @abstractmethod
    def getCost(self) -> int:
        pass

    @abstractmethod
    def getDescription(self) -> str:
        pass


class SimpleCoffee(Coffee):
    def getCost(self) -> int:
        return 10

    def getDescription(self) -> str:
        return "Simple coffee"


class MilkCoffee(Coffee):
    def __init__(self, coffee: Coffee) -> None:
        self.coffee = coffee

    def getCost(self):
        return self.coffee.getCost() + 2

    def getDescription(self):
        return self.coffee.getDescription() + '、ミルク'


class WhipCoffee(Coffee):
    def __init__(self, coffee: Coffee) -> None:
        self.coffee = coffee

    def getCost(self):
        return self.coffee.getCost() + 5

    def getDescription(self):
        return self.coffee.getDescription() + '、ホイップ'


class VanillaCoffee(Coffee):
    def __init__(self, coffee: Coffee) -> None:
        self.coffee = coffee

    def getCost(self):
        return self.coffee.getCost() + 5

    def getDescription(self):
        return self.coffee.getDescription() + '、バニラ'


someCoffee = SimpleCoffee()
print(someCoffee.getCost())
print(someCoffee.getDescription())

someCoffee = MilkCoffee(someCoffee)
print(someCoffee.getCost())
print(someCoffee.getDescription())

someCoffee = WhipCoffee(someCoffee)
print(someCoffee.getCost())
print(someCoffee.getDescription())

someCoffee = VanillaCoffee(someCoffee)
print(someCoffee.getCost())
print(someCoffee.getDescription())

File: draft_design_pattern/decorator.ts
export {};

interface Coffee {
  getCost(): number;
  getDescription(): string;
}

class SimpleCoffee implements Coffee {
  getCost(): number {
    return 10;
  }
  getDescription(): string {
    return "Simple coffee";
  }
}

class MilkCoffee implements Coffee {
  constructor(private coffee: Coffee) {}
  getCost(): number {
    return this.coffee.getCost() + 2;
  }
  getDescription(): string {
    return this.coffee.getDescription() + "、ミルク";
  }
}

class WhipCoffee implements Coffee {
  constructor(private coffee: Coffee) {}
  getCost(): number {
    return this.coffee.getCost() + 5;
  }
  getDescription(): string {
    return this.coffee.getDescription() + "、ホイップ";
  }
}

class VanillaCoffee implements Coffee {
  constructor(private coffee: Coffee) {}
  getCost(): number {
    return this.coffee.getCost() + 5;
  }
  getDescription(): string {
    return this.coffee.getDescription() + "、バニラ";
  }
}

let someCoffee = new SimpleCoffee();
console.log(someCoffee.getCost());
console.log(someCoffee.getDescription());

someCoffee = new MilkCoffee(someCoffee);
console.log(someCoffee.getCost());
console.log(someCoffee.getDescription());

someCoffee = new WhipCoffee(someCoffee);
console.log(someCoffee.getCost());
console.log(someCoffee.getDescription());

someCoffee = new VanillaCoffee(someCoffee);
console.log(someCoffee.getCost());
console.log(someCoffee.getDescription());
